torchsde/_core/sdeint.py

```diff
@@ -145,63 +146,103 @@ def check_contract(sde, y0, ts, bm, method, names, logqp):
+            state_sizes.append(shape[1])
+            noise_sizes.append(shape[2])
+
+    has_f = False
+    has_g = False
+    if hasattr(sde, 'f'):
+        has_f = True
+        f_drift_shape = tuple(sde.f(ts[0], y0).size())
+        _check_2d('Drift', f_drift_shape)
+    if hasattr(sde, 'g'):
+        has_g = True
+        g_diffusion_shape = tuple(sde.g(ts[0], y0).size())
+        _check_2d_or_3d('Diffusion', g_diffusion_shape)
+    if hasattr(sde, 'f_and_g'):
+        has_f = True
+        has_g = True
+        _f, _g = sde.f_and_g(ts[0], y0)
+        f_drift_shape = tuple(_f.size())
+        g_diffusion_shape = tuple(_g.size())
+        _check_2d('Drift', f_drift_shape)
+        _check_2d_or_3d('Diffusion', g_diffusion_shape)
+    if hasattr(sde, 'g_prod'):
+        has_g = True
+        if len(noise_sizes) == 0:
+            raise ValueError("Cannot infer noise size (i.e. number of Brownian motion channels). Either pass `bm` "
+                             "explicitly, or specify one of the `g`, `f_and_g` functions.`")
+        v = torch.randn(batch_sizes[0], noise_sizes[0], dtype=y0.dtype, device=y0.device)
+        g_prod_shape = tuple(sde.g_prod(ts[0], y0, v).size())
+        _check_2d('Diffusion-vector product', g_prod_shape)
+    if hasattr(sde, 'f_and_g_prod'):
+        has_f = True
+        has_g = True
+        if len(noise_sizes) == 0:
+            raise ValueError("Cannot infer noise size (i.e. number of Brownian motion channels). Either pass `bm` "
+                             "explicitly, or specify one of the `g`, `f_and_g` functions.`")
+        v = torch.randn(batch_sizes[0], noise_sizes[0], dtype=y0.dtype, device=y0.device)
+        _f, _g_prod = sde.f_and_g_prod(ts[0], y0, v)
+        f_drift_shape = tuple(_f.size())
+        g_prod_shape = tuple(_g_prod.size())
+        _check_2d('Drift', f_drift_shape)
+        _check_2d('Diffusion-vector product', g_prod_shape)
+
+    if not has_f:
+        raise ValueError("sde must define at least one of `f`, `f_and_g`, or `f_and_g_prod`. (Or possibly more "
+                         "depending on the method chosen.)")
+    if not has_g:
+        raise ValueError("sde must define at least one of `g`, `f_and_g`, `g_prod` or `f_and_g_prod`. (Or possibly "
+                         "more depending on the method chosen.)")
+
+    for batch_size in batch_sizes[1:]:
+        if batch_size != batch_sizes[0]:
+            raise ValueError("Batch sizes not consistent.")
+    for state_size in state_sizes[1:]:
+        if state_size != state_sizes[0]:
+            raise ValueError("State sizes not consistent.")
+    for noise_size in noise_sizes[1:]:
+        if noise_size != noise_sizes[0]:
+            raise ValueError("Noise sizes not consistent.")
+
+    if sde.noise_type == NOISE_TYPES.scalar:
+        if noise_sizes[0] != 1:
+            raise ValueError(f"Scalar noise must have only one channel; the diffusion has {noise_sizes[0]} noise "
+                             f"channels.")
+
```

Comment:
This red+green block above here is the entirety of bullet point 4, on better contract checking. It's actually in this PR because it interacts with the `f_and_g` functionality, as we need to check that what that returns is sane, so I don't think it makes sense to split this into a separate PR.

I'd note that about half of the changes apparently in this PR are actually #85, which makes this PR look bigger than it is. Once that's in it'll be easier to read.